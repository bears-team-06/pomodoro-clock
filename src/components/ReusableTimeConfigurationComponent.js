import React, { Component } from 'react';

class ReusableTimeConfigurationComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {
            time: this.props.startValue
        }
    }

    onIncrementButtonClick = () => {
        this.setState(prevState => {
            return {time: prevState.time + this.props.minimumChange}
        })
    };

    onDecrementButtonClick = () => {
        this.setState(prevState => {
            return {time: prevState.time - this.props.minimumChange}
        })
    };

    get incrementButtonDisabled() {
        return this.state.time >= this.props.maximumBreakLength;
    }

    get decrementButtonDisabled() {
        return this.state.time <= this.props.minimumBreakLength;
    }

    get timeString() {
        let minutes = Math.floor(this.state.time / 60);
        return minutes > 9 ? minutes : ` ${minutes}`
    }

    render() {
        return (
            <div>
                <span id={'label'}>{this.props.labelName}</span>
                <span><button id='decrement' onClick={this.onDecrementButtonClick} disabled={this.decrementButtonDisabled}>-</button></span>
                <span><span id={'time'}>{this.timeString}</span></span>
                <span><button id='increment' onClick={this.onIncrementButtonClick} disabled={this.incrementButtonDisabled}>+</button></span>
            </div>
        );
    }
}

export default ReusableTimeConfigurationComponent;