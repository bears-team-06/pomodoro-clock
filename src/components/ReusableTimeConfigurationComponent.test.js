import ReusableTimeConfigurationComponent from "./ReusableTimeConfigurationComponent";
import { shallow } from "enzyme";
import React from "react";

describe("ReusableTimeConfigurationComponent", () => {
  describe("when given certain properties", () => {
    it("matches snapshot", () => {
      const timeConfigurationComponent = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={300}
          minimumChange={60}
          maximumLength={3600}
          minimumLength={60}
        />
      );
      expect(timeConfigurationComponent).toMatchSnapshot();
    });

    it("displays correct start time and label", () => {
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={300}
          minimumChange={60}
          maximumLength={3600}
          minimumLength={60}
        />
      );
      const label = wrapper.find("#label");
      expect(label.text()).toBe("Break Length");
      expect(wrapper.find("#time").text().trim()).toBe("5");
      expect(wrapper.find("#increment").prop("disabled")).toBe(false);
      expect(wrapper.find("#decrement").prop("disabled")).toBe(false);
    });
  });

  describe("when pressing increment button", () => {
    it("on Change is called with incremented Value", () => {
      let newTime = null;
      let onChange = newValue => {
        newTime = newValue;
      };
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={300}
          minimumChange={60}
          maximumLength={3600}
          minimumLength={60}
          onChange={onChange}
        />
      );

      wrapper.find("#increment").simulate("click");
      expect(newTime).toBe(360);
    });
  });

  describe("when pressing decrement button", () => {
    it("on Change is called with decremented value", () => {
      let newTime = null;
      let onChange = newValue => {
        newTime = newValue;
      };
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={300}
          minimumChange={60}
          maximumLength={3600}
          minimumLength={60}
          onChange={onChange}
        />
      );
      wrapper.find("#decrement").simulate("click");
      expect(newTime).toBe(240);
    });
  });

  describe("when timeLength is equal or greater than maximumLength", () => {
    it("disables increment button", () => {
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={420}
          minimumChange={60}
          maximumLength={360}
          minimumLength={60}
        />
      );

      expect(
        wrapper
          .find("#time")
          .text()
          .trim()
      ).toBe("7");
      expect(wrapper.find("#increment").prop("disabled")).toBe(true);
    });
  });

  describe("when timeLength is less than or equal to minimumChange", () => {
    it("enables decrement button", () => {
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={60}
          minimumChange={60}
          maximumLength={3600}
          minimumLength={60}
        />
      );
      expect(
        wrapper
          .find("#time")
          .text()
          .trim()
      ).toBe("1");
      expect(wrapper.find("#decrement").prop("disabled")).toBe(true);
    });
  });

  describe("when time length is greater than minimum length and less than maximum length", () => {
    it("enables both increment and decrement buttom", () => {
      const wrapper = shallow(
        <ReusableTimeConfigurationComponent
          labelName="Break Length"
          timeLength={120}
          minimumChange={60}
          maximumLength={360}
          minimumLength={60}
        />
      );
      expect(wrapper.find("#increment").prop("disabled")).toBe(false);
      expect(wrapper.find("#decrement").prop("disabled")).toBe(false);
    });
  });
});
