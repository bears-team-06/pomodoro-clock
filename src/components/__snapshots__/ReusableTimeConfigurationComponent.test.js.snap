// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReusableTimeConfigurationComponent when given certain properties matches snapshot 1`] = `
<div
  className="length-configuration"
>
  <span
    className="label"
  >
    Break Length
  </span>
  <button
    className="decrement btn rounded-button"
    onClick={[Function]}
  >
    <span>
      -
    </span>
  </button>
  <span
    className="time"
  >
     5
  </span>
  <button
    className="increment btn rounded-button"
    onClick={[Function]}
  >
    <span>
      +
    </span>
  </button>
</div>
`;
